hypothesis_regex.py

```diff
@@ -1,102 +1,335 @@
+        Add unicode category to set
+
+        Unicode categories are strings like 'Ll', 'Lu', 'Nd', etc.
+        See `unicodedata.category()`
+        '''
+        if category == sre.CATEGORY_DIGIT:
+            self._categories |= UNICODE_DIGIT_CATEGORIES
+        elif category == sre.CATEGORY_NOT_DIGIT:
+            self._categories |= UNICODE_CATEGORIES - UNICODE_DIGIT_CATEGORIES
+        elif category == sre.CATEGORY_SPACE:
+            self._categories |= UNICODE_SPACE_CATEGORIES
+            for c in (UNICODE_SPACE_CHARS if self._unicode else SPACE_CHARS):
+                self._whitelist_chars.add(c)
+        elif category == sre.CATEGORY_NOT_SPACE:
+            self._categories |= UNICODE_CATEGORIES - UNICODE_SPACE_CATEGORIES
+            for c in (UNICODE_SPACE_CHARS if self._unicode else SPACE_CHARS):
+                self._blacklist_chars.add(c)
+        elif category == sre.CATEGORY_WORD:
+            self._categories |= UNICODE_WORD_CATEGORIES
+            self._whitelist_chars.add(u'_')
+            if HAS_WEIRD_WORD_CHARS and self._unicode:
+                for c in UNICODE_WEIRD_NONWORD_CHARS:
+                    self._blacklist_chars.add(c)
+        elif category == sre.CATEGORY_NOT_WORD:
+            self._categories |= UNICODE_CATEGORIES - UNICODE_WORD_CATEGORIES
+            self._blacklist_chars.add(u'_')
+            if HAS_WEIRD_WORD_CHARS and self._unicode:
+                for c in UNICODE_WEIRD_NONWORD_CHARS:
+                    self._whitelist_chars.add(c)
+
+    def add_chars(self, chars):
+        'Add given chars to char set'
+        for c in chars:
+            if self._ignorecase:
+                self._whitelist_chars.add(c.lower())
+                self._whitelist_chars.add(c.upper())
+            else:
+                self._whitelist_chars.add(c)
 
 
 @hs.defines_strategy
 def regex(regex):
-    """Return strategy that generates strings that match given regex."""
-    if hasattr(regex, 'pattern'):
-        regex = regex.pattern
+    """Return strategy that generates strings that match given regex.
+
+    Regex can be either a string or compiled regex (through `re.compile()`).
+
+    You can use regex flags (such as `re.IGNORECASE`, `re.DOTALL` or `re.UNICODE`)
+    to control generation. Flags can be passed either in compiled regex (specify
+    flags in call to `re.compile()`) or inside pattern with (?iLmsux) group.
+
+    Some tricky regular expressions are partly supported or not supported at all.
+    "^" and "$" do not affect generation. Positive lookahead/lookbehind groups
+    are considered normal groups. Negative lookahead/lookbehind groups do not do
+    anything. Ternary regex groups ('(?(name)yes-pattern|no-pattern)') are not
+    supported at all.
+    """
+    if not hasattr(regex, 'pattern'):
+        regex = re.compile(regex)
+
+    pattern = regex.pattern
+    flags = regex.flags
+
+    codes = sre.parse(pattern)
+
+    return _strategy(codes, Context(flags=flags)).filter(regex.match)
```

Comment:
There are some cases, which are either hard to ensure are generated correctly, nor to restructure to maintain invariants. E.g. ".*(?<!foo)bar$" (match anything that ends with "bar" unless "bar" is preceeded with "foo"). It is hard to accomplish that with using `filter()` somewhere in strategy (you could say "hey, let's get whatever strategies preceed that and then filter", but is almost the same as filtering the whole expression although whole expression filter is easier to implement).

My preference is to do best effort generating stuff without putting complex logic of rejecting regexes that in some cases we can not reliably generate and let people and/or health checks to judge.